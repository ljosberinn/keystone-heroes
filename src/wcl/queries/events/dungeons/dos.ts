import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  BeginCastEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const DOS_URN = 228_626 as const;
export const DOS_URN_OPENING = 340_004 as const;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=2&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20228626
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 34007, endTime: 1759795, filterExpression: "type = \"applydebuff\" and target.type = \"npc\" and ability.id = 228626") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "npc" and ability.id = ${DOS_URN}`,
  `ability.id = ${DOS_URN_OPENING}`,
];

const isDosUrnEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: DOS_URN,
});

const isDosUrnBeginCastEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: DOS_URN_OPENING,
});

export const getDOSEvents = (
  allEvents: AllTrackedEventTypes[]
): (ApplyDebuffEvent | BeginCastEvent)[] => {
  return [
    ...allEvents.filter(isDosUrnEvent),
    ...allEvents.filter(isDosUrnBeginCastEvent),
  ];
};
