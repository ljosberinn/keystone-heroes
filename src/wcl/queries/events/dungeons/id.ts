import type { AllTrackedEventTypes, DamageEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const IRON_DOCKS_CRUSHED = 167_862 as const;

export const filterExpression = [
  `type = "damage" and ability.id = ${IRON_DOCKS_CRUSHED}`,
];

const isIronStarCrushedEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: IRON_DOCKS_CRUSHED,
});

export const getIDEvents = (
  allEvents: AllTrackedEventTypes[]
): DamageEvent[] => {
  return allEvents.filter(isIronStarCrushedEvent);
};
