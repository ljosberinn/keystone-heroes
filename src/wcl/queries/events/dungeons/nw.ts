import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  HealEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const NW = {
  ORB: 328_406,
  HAMMER: 328_128,
  SPEAR: 328_351,
  KYRIAN_ORB_BUFF: 335_161,
  KYRIAN_ORB_HEAL: 344_422,
  KYRIAN_ORB_DAMAGE: 344_421,
} as const;

/**
 * @see https://www.warcraftlogs.com/reports/Jq7KrbYV1hmTWMyw#fight=4&view=events&pins=2%24Off%24%23244F4B%24expression%24(type%20%3D%20%22heal%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20344422)%0Aor%20(type%20%3D%20%22damage%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20in%20(328128,%20344421,%20328351,%20328406))%0Aor%20((type%20%3D%20%22begincast%22%20or%20type%20%3D%20%22cast%22)%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20in%20(328128,%20328351))%0Aor%20(type%20%3D%20%22applybuff%22%20and%20source.type%20%3D%20%22player%22%20and%20ability.id%20%3D%20335161)
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "Jq7KrbYV1hmTWMyw") {
 *       fights(fightIDs: [4]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 1764363, endTime: 3716117, filterExpression: "(type = \"heal\" and source.type = \"player\" and ability.id = 344422) or (type = \"damage\" and source.type = \"player\" and ability.id in (328406, 328128, 328351, 344421))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applybuff" and ability.id = ${NW.KYRIAN_ORB_BUFF}`,
  `type = "heal" and source.type = "player" and ability.id = ${NW.KYRIAN_ORB_HEAL}`,
  `type = "damage" and source.type = "player" and ability.id in (${[
    NW.HAMMER,
    NW.KYRIAN_ORB_DAMAGE,
    NW.SPEAR,
    NW.ORB,
  ].join(", ")})`,
  `(type = "begincast" or type = "cast") and source.type = "player" and ability.id in (${[
    NW.HAMMER,
    NW.SPEAR,
  ].join(", ")})`,
];

const isNwSpearEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.SPEAR,
});

const isNwSpearCastEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: NW.SPEAR,
});

const isNwSpearBeginCastEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: NW.SPEAR,
});

const isNwHammerEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.HAMMER,
});

const isNwHammerCastEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: NW.HAMMER,
});

const isNwHammerBeginCastEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: NW.HAMMER,
});

const isNwOrbEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.ORB,
});

const isNwOrbCastEvent = createIsSpecificEvent<CastEvent>({
  type: "cast",
  abilityGameID: NW.ORB,
});

const isNwKyrianOrbDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: NW.KYRIAN_ORB_DAMAGE,
});

const isNwKyrianOrbHealEvent = createIsSpecificEvent<HealEvent>({
  type: "heal",
  abilityGameID: NW.KYRIAN_ORB_HEAL,
});

const isNwKyrianOrbApplyBuffEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: NW.KYRIAN_ORB_BUFF,
});

export const getNWEvents = (
  allEvents: AllTrackedEventTypes[]
): (
  | ApplyBuffEvent
  | DamageEvent
  | HealEvent
  | BeginCastEvent
  | CastEvent
)[] => {
  const spearDamageEvents = allEvents.filter(isNwSpearEvent);

  const spearBeginCastEvents = allEvents.filter(isNwSpearBeginCastEvent);
  const spearCastEvents = allEvents.filter(isNwSpearCastEvent);

  const hammerBeginCastEvents = allEvents.filter(isNwHammerBeginCastEvent);
  const hammerCastEvents = allEvents.filter(isNwHammerCastEvent);
  const hammerDamageEvents = allEvents.filter(isNwHammerEvent);

  const orbCastEvents = allEvents.filter(isNwOrbCastEvent);
  const orbDamageEvents = allEvents.filter(isNwOrbEvent);

  const kyrianOrbApplyBuffEvents = allEvents.filter(
    isNwKyrianOrbApplyBuffEvent
  );
  const kyrianOrbDamageEvents = allEvents.filter(isNwKyrianOrbDamageEvent);
  const kyrianOrbHealEvents = allEvents.filter(isNwKyrianOrbHealEvent);

  return [
    ...spearDamageEvents,
    ...spearBeginCastEvents,
    ...spearCastEvents,

    ...hammerBeginCastEvents,
    ...hammerCastEvents,
    ...hammerDamageEvents,

    ...orbCastEvents,
    ...orbDamageEvents,

    ...kyrianOrbApplyBuffEvents,
    ...kyrianOrbDamageEvents,
    ...kyrianOrbHealEvents,
  ];
};
