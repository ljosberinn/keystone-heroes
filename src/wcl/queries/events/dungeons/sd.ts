import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  BeginCastEvent,
  CastEvent,
  RemoveBuffEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const SD_LANTERN_OPENING = 340_013 as const;
export const SD_LANTERN_BUFF = 340_433 as const;

/**
 * @see https://www.warcraftlogs.com/reports/mQn47kFCbpThBtrG#fight=15&type=summary&view=events&pins=2%24Off%24%23909049%24expression%24ability.id%20%3D%20340433%20and%20type%20in%20(%22applybuff%22,%20%22applybuffstack%22,%20%22removebuff%22)%5E2%24Off%24%23909049%24expression%24ability.id%20%3D%20340013%20and%20type%20%3D%22begincast%22
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "mQn47kFCbpThBtrG") {
 *       fights(fightIDs: [15]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 11536849, endTime: 13848195, filterExpression: "(type = \"begincast\" and ability.id = 340013) or (ability.id = 340433 and type in (\"applybuff\", \"applybuffstack\", \"removebuff\"))") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
const lanternOpeningExpression = `type = "begincast" and ability.id = ${SD_LANTERN_OPENING}`;
const lanternBuffExpression = `ability.id = ${SD_LANTERN_BUFF} and type in ("applybuff", "applybuffstack", "removebuff")`;

export const filterExpression = [
  lanternOpeningExpression,
  lanternBuffExpression,
];

const isSdLanternOpeningEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: SD_LANTERN_OPENING,
});

const isSdLanternBuffEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: SD_LANTERN_BUFF,
});

const isSdLanternRemoveBuffEvent = createIsSpecificEvent<RemoveBuffEvent>({
  type: "removebuff",
  abilityGameID: SD_LANTERN_BUFF,
});

const isSdLanternApplyBuffStackEvent =
  createIsSpecificEvent<ApplyBuffStackEvent>({
    type: "applybuffstack",
    abilityGameID: SD_LANTERN_BUFF,
  });

export const getSDEvents = (
  allEvents: AllTrackedEventTypes[]
): (
  | BeginCastEvent
  | ApplyBuffEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
  | CastEvent
)[] => {
  const partial = [
    ...allEvents.filter(isSdLanternOpeningEvent),
    ...allEvents.filter(isSdLanternBuffEvent),
    ...allEvents.filter(isSdLanternApplyBuffStackEvent),
  ];

  // experimental detection
  const castEvents = partial
    .sort((a, b) => a.timestamp - b.timestamp)
    .reduce<CastEvent[]>((acc, event, index, arr) => {
      // find the last begincast event before a different type occurs
      if (event.type === "begincast" && arr[index + 1]?.type !== "begincast") {
        return [
          ...acc,
          {
            ...event,
            type: "cast",
            timestamp: event.timestamp + 3200,
          },
        ];
      }

      return acc;
    }, []);

  return [
    ...allEvents.filter(isSdLanternRemoveBuffEvent),
    ...partial,
    ...castEvents,
  ];
};
