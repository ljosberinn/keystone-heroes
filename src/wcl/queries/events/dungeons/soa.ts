import type { AllTrackedEventTypes, ApplyDebuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const SOA_SPEAR = 339_917 as const;

/**
 * @see https://www.warcraftlogs.com/reports/JVnw7NF6hyz8QqX4#fight=20&type=summary&view=events&pins=2%24Off%24%23a04D8A%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20339917&hostility=1
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "JVnw7NF6hyz8QqX4") {
 *       fights(fightIDs: [20]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 11052027, endTime: 13152414, filterExpression: "type = \"applydebuff\" and ability.id = 339917") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and ability.id = ${SOA_SPEAR}`,
];

const isSoaSpearEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: SOA_SPEAR,
});

export const getSOAEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return allEvents.filter(isSoaSpearEvent);
};
