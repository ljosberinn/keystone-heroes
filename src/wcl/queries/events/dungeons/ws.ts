import type { AllTrackedEventTypes, DamageEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const WS = {
  ROCKET_BARRAGE: 294_128,
  ANTI_PERSONNEL_SQUIRREL: 293_861,
} as const;

export const filterExpression = [
  `type = "damage" and ability.id in (${Object.values(WS).join(", ")})`,
];

const isWorkshopDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: Object.values(WS),
});

export const getWSEvents = (
  allEvents: AllTrackedEventTypes[]
): DamageEvent[] => {
  return allEvents.filter(isWorkshopDamageEvent);
};
