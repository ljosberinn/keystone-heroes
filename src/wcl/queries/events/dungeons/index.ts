import { DungeonIDs } from "../../../../db/data/dungeons";
import type { AllTrackedEventTypes } from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import {
  filterExpression as motsFilterExpression,
  getMOTSEvents,
} from "./mots";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

const dungeonExpressionMap: Record<DungeonIDs, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATER_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
  [DungeonIDs.MISTS_OF_TIRNA_SCITHE]: motsFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  dungeonExpressionMap[id];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes[],
  dungeonID: DungeonIDs,
  actorIDSet: Set<number>
): ReturnType<
  | typeof getDOSEvents
  | typeof getHOAEvents
  | typeof getPFEvents
  | typeof getSDEvents
  | typeof getSOAEvents
  | typeof getTOPEvents
  | typeof getNWEvents
  | typeof getMOTSEvents
> => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, actorIDSet);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATER_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    case DungeonIDs.MISTS_OF_TIRNA_SCITHE:
      return getMOTSEvents(allEvents);
    default:
      return [];
  }
};
