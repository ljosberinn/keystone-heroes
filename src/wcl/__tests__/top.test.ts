import { getTOPEvents } from "../queries/events/dungeons/top";
import allEvents from "./fixtures/allEvents.json";

describe("getTOPEvents", () => {
  test("works", () => {
    expect(getTOPEvents(allEvents)).toMatchSnapshot();
  });
});
