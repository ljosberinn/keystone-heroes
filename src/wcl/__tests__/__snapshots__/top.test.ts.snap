// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTOPEvents works 1`] = `
Array [
  Object {
    "abilityGameID": 340378,
    "pin": "0",
    "sourceID": 105,
    "targetID": 105,
    "timestamp": 5032705,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340378,
    "pin": "0",
    "sourceID": 105,
    "targetID": 105,
    "timestamp": 5549610,
    "type": "applybuff",
  },
  Object {
    "abilityGameID": 340378,
    "pin": "0",
    "sourceID": 105,
    "targetID": 105,
    "timestamp": 6031023,
    "type": "applybuff",
  },
]
`;
